build.sh

```diff
@@ -0,0 +1,159 @@
+fi
+
+[[ ${CLEAN} -eq 1 ]] && rm -rf "${OE_SEC_BUILD}"
+[[ ! -d "${OE_SEC_BUILD}" ]] && mkdir -p "${OE_SEC_BUILD}"
+
+if [[ ! -d "${OE_SEC_TOOLS}" ]]; then
+    mkdir -p "${OE_SEC_TOOLS}"
+    pushd "${OE_SEC_TOOLS}"
+    wget "${OE_LLVM_URL}"
+    unzip oe-llvm-1.0.zip
+    popd
+fi
+
+MAKE_THREADS=$(nproc)
+
+# Building debug version of Intel PSW.
+if [[ ${BUILD_INTEL_SGX_PSW} -eq 1 ]]; then
+    if [[ ! -d "/opt/intel/sgxsdk" ]]; then
+        pushd "${OE_SEC_TOOLS}"
+        wget "${INTEL_SGX_SDK}"
+        chmod +x ./"${INTEL_SGX_SDK_PACKAGE}"
+        sudo ./"${INTEL_SGX_SDK_PACKAGE}" <<EOF
+no
+/opt/intel
+EOF
+        popd
+    fi
+
+    pushd "${OE_3RDPARTY}"
+    make -C linux-sgx clean all
+    make -C linux-sgx preparation
+    make -C linux-sgx psw DEBUG=1 -j ${MAKE_THREADS}
+    popd
+fi
+
+# Building OE in release mode to build an enclave which will used as corpus data in create api fuzzing.
+if [[ ! -d "${OE_UNINSTRUMENTED_BUILD}" ]]; then
+    mkdir -p "${OE_UNINSTRUMENTED_BUILD}"
+    mkdir -p "${OE_UNINSTRUMENTED_INSTALL_PREFIX}"
+    pushd "${OE_UNINSTRUMENTED_BUILD}"
+    cmake "${OE_SRC_ROOT}" -GNinja \
+        -DBUILD_TESTS=OFF \
+        -DCMAKE_C_COMPILER="${CLANG}" \
+        -DCMAKE_CXX_COMPILER="${CLANG_CPP}" \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DUSE_DEBUG_MALLOC=OFF \
+        -DCMAKE_INSTALL_PREFIX="${OE_UNINSTRUMENTED_INSTALL_PREFIX}"
+    ninja install -j ${MAKE_THREADS}
+    popd
+fi
+
+# Building OE in debug mode and instrumented with OE-LLVM sanitizers.
+if [[ ! -d "${OE_INSTRUMENTED_BUILD}" ]]; then
+    mkdir -p "${OE_INSTRUMENTED_BUILD}"
+    mkdir -p "${OE_INSTRUMENTED_INSTALL_PREFIX}"
+    pushd "${OE_INSTRUMENTED_BUILD}"
+    cmake "${OE_SRC_ROOT}" -GNinja \
+        -DENABLE_FUZZING=ON \
+        -DBUILD_OEGENERATE_TOOL=OFF \
+        -DBUILD_TESTS=OFF \
+        -DCMAKE_C_COMPILER="${CLANG}" \
+        -DCMAKE_CXX_COMPILER="${CLANG_CPP}" \
+        -DCMAKE_BUILD_TYPE=Debug \
+        -DCMAKE_C_FLAGS_DEBUG="-O0 -g" \
+        -DCMAKE_CXX_FLAGS_DEBUG="-O0 -g" \
+        -DUSE_DEBUG_MALLOC=OFF \
+        -DCMAKE_INSTALL_PREFIX="${OE_INSTRUMENTED_INSTALL_PREFIX}"
+    ninja install -j ${MAKE_THREADS}
+    popd
+fi
+
+# Building an enclave with un-instrumented OE-SDK.
+rm -rf "${OE_SEC_CORPUS_BUILD}"
+mkdir -p "${OE_SEC_CORPUS_BUILD}"
+pushd "${OE_SEC_CORPUS_BUILD}"
+cmake "${OE_SEC_ENCLAVE_CORPUS}" -GNinja \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_C_COMPILER="${CLANG}" \
+    -DCMAKE_CXX_COMPILER="${CLANG_CPP}" \
+    -DCMAKE_PREFIX_PATH="${OE_UNINSTRUMENTED_INSTALL_PREFIX}"
+ninja -j ${MAKE_THREADS}
+popd
+
+# Building fuzzer targets with instrumented OE-SDK.
+rm -rf "${OE_SEC_FUZZING_BUILD}"
+mkdir -p "${OE_SEC_FUZZING_BUILD}"
+pushd "${OE_SEC_FUZZING_BUILD}"
+cmake "${OE_SEC_ROOT}" -GNinja \
+    -DCMAKE_BUILD_TYPE=Debug \
+    -DCMAKE_C_FLAGS_DEBUG="-O0 -g" \
+    -DCMAKE_CXX_FLAGS_DEBUG="-O0 -g" \
```

Comment:
Shouldn't the fuzzer targets be built with the expected optimization level?

Reply:
Optimization levels has to be disabled for LLVM sanitizers to instrument the edges and to insert callbacks in IR pass.

Reply:
This is unfortunate given that optimizations tend to remove some security protections.